src/Portal.tsx

```diff
@@ -14,44 +20,39 @@ export interface PortalProps {
 const Portal = forwardRef<PortalRef, PortalProps>((props, ref) => {
   const { didUpdate, getContainer, children } = props;
 
-  const parentRef = useRef<ParentNode>();
-  const containerRef = useRef<HTMLElement>();
+  const parentMap = useMemo(() => new WeakMap<HTMLElement, ParentNode>(), []);
+  const [portalContainer, setPortalContainer] = useState<HTMLElement>();
 
   // Ref return nothing, only for wrapper check exist
   useImperativeHandle(ref, () => ({}));
 
-  // Create container in client side with sync to avoid useEffect not get ref
-  const initRef = useRef(false);
-  if (!initRef.current && canUseDom()) {
-    containerRef.current = getContainer();
-    parentRef.current = containerRef.current.parentNode;
-    initRef.current = true;
-  }
-
   // [Legacy] Used by `rc-trigger`
   useEffect(() => {
     didUpdate?.(props);
   });
 
-  useEffect(() => {
+  useLayoutEffect(() => {
+    const container = getContainer();
```

Comment:
do not call `getContainer` in effect since some developer has some code in `getContainer` as sync logic. We should not break that.

Reply:
Do you have any suggestions?

Reply:
why this ref mark not work?
https://github.com/react-component/util/blob/master/src/Portal.tsx#L25

Reply:
Function component bodies will be intentionally double-invoked, `ref` is independent each time, and the first time will not trigger _effects_ to ensure the body is PURE.

See [detecting unexpected side effects](https://reactjs.org/docs/strict-mode.html#detecting-unexpected-side-effects)

Reply:
As you can see, the first DOM (0.17325491913791602) was created, but will not be destroyed.

Reply:
Tested memo as the same result: https://stackblitz.com/edit/react-mawziy?file=index.js

Reply:
headache. Seems no to break `rc-util` exist usage. We have to make getContainer unique on rc-trigger side

Reply:
Making `getContainer` unique works, but it's just a tradeoff. The `Portal` should be side-effects free in the long term. How do u think?

Reply:
Yes. But we still need not break current user behavior. `getContainer` is currently best way.

Reply:
Okay, could you please review this PR? https://github.com/react-component/trigger/pull/292/files